### DISCUSSION TITLE
MEMS microphone on sensor board via I2C

### DISCUSSION THREAD

**Discussion by mwebr** [joeycastillo/Sensor-Watch#199] [Q&A]
Hi everyone,

First of all: this is a great project and I am having a lot of fun with programming the watch for my needs!
Currently, I am working on a topic that involves the integration of a MEMS microphone onto the sensor I2C extension of the sensor watch board for temporary use (analyzing some sound patterns). 

I am seeking your expertise and advice in this matter. I would be grateful if you could share any ideas or suggestions.

I am stuck on how to implement the microphone. I assume, that an analog MEMS microphone wouldn't be a good choice, therefore I am planning to integrate the Knowles SPH0645LM4H I2C MEMS.

Some questions on that (I am quite a newbie on that concern):
- Does it make sense concerning power consumption? I have seen, that it has a sleep mode.
- Does it fit from the height (1,2 mm)?
- What supply voltage comes out of the sensor board?
- Does the software integration work similar to the accelerometer example?

I would also contribute my work to the repo, if desired.

Best regards!

**Comment by joeycastillo** [comment]
This is a great idea for a sensor board, and I'm stoked that you're interested in doing it! I have thought about it and I have a few notes. First, the mic you're describing is I2S, which is different than I2C, and there may be some challenges to making it work. But generally speaking, to answer your questions: 

* Concerning power consumption, if there's a sleep mode under 10µA, it should be workable. This one is 3µA; good!
* Height: anything under 1mm should be an easy fit, and you _might_ even be able to fit something 1.1 mm tall in there. But this looks like it's 0.98 mm tall which is fine!
* Supply voltage to a sensor board is between 3.1 volts (full battery) and 2.7 volts (after a year or more). Beyond that, you'll have a few more months where it sags from 2.7 to 2 volts before the battery dies.
* Software integration for this one would be wildly different from the accelerometer example, and I think there are two approaches you could use, outlined below.

First off, I think using an analog MEMS microphone instead could be a great choice, though I don't know that for sure. The SAM L22 has an integrated ADC capable of 1 million samples per second at 12-bit resolution, so you could probably get useful analog data from the ADC peripheral, and even set up DMA for this purpose. That's one approach. 

With the I2S microphone, I think the challenge is going to be this: **the SAM L22 has no dedicated I2S peripheral**. That's a bummer, and on its face suggests you might not be able to use it — but I think there may still be a way. The gist is that I2S is [sort of similar to SPI](https://learn.adafruit.com/adafruit-i2s-mems-microphone-breakout?view=all#i2s-data-pins-2252707) in that there's a clock line and a data line. The difference is that there's also a word select line, which controls which channel of microphone is transmitting. With some extreme software cleverness, I think you _might_ be able to hack something together with a sensor board pinout as follows:

* Microphone BCLK connected to A1 (SPI SCK)
* Microphone WS connected to A2 (SPI MOSI / CCL 0 out)
* Microphone DATA connected to A4 (SPI MISO)
* Microphone SEL tied to either VCC or GND

I sense (no guarantees!) that you could either use the SPI MOSI line to drive the word select line high and low, OR MAYBE you could use the SAM L22's [Configurable Custom Logic](https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-42448-SAM-Configurable-Custom-Logic-CCL-Driver_AT10942_Application%20Note.pdf) driver to hack together an output that toggles with the correct timing. There's an interesting TI app note about [interfacing an I2S device to an MSP430 using SPI](https://www.ti.com/lit/an/slaa449a/slaa449a.pdf) and a couple of logic gates to turn the BCLK signal into a WS or left/right clock. If you could implement these logic gates using the CCL peripheral, you might be able to get away with a sensor board containing nothing but the microphone and a decoupling capacitor.

Hope this helps!
